core/include/scipp/core/element_array.h

```diff
@@ -12,9 +12,9 @@
 
 namespace scipp::core {
 
-/// Replacement for C++20 std::make_unique_default_init
+/// Create a unique pointer to an array of default initialized elements.
 template <class T> auto make_unique_default_init(const scipp::index size) {
-  return std::unique_ptr<T>(new std::remove_extent_t<T>[size]);
+  return std::unique_ptr<T>(new std::remove_extent_t<T>[size]());
```

Comment:
Not sure I understand correctly, but according to https://en.cppreference.com/w/cpp/language/default_initialization the `()` was needed until C++03 (the article leaves it unclear what that `()` does now?)? Can you elaborate what the difference is?

Reply:
See item 5 here: https://en.cppreference.com/w/cpp/memory/unique_ptr/make_unique. No `()`. With `()` you get item 2, which is value-init, not default-init.

Reply:
On the first page under 'Explanation':
> otherwise, nothing is done: the objects with automatic storage duration (and their subobjects) are initialized to indeterminate values. 

This bullet applies to fundamental types.

You are correct in that the change does not fit the name. Default initialisation means that ints / floats remain uninitialised. Did I misunderstand the meaning of the `default_init_elements_t` tag? `element_array::resize(size)` seems to value-init its elements and `element_array(size, default_init_elements)` leaves them uninitialised.

Reply:
Yes, default-init means uninitialized, and that is exactly what we want here. It is required, e.g., in `transform` to create the output, which is subsequently overwritten. If we would zero-initialize it there would be two writes, degrading performance of most non-in-place operations.

`std::make_unique_default_init` was apparently the original name. I suppose you were not the only one who was confused, so the standard now calls it `std::make_unique_for_overwrite`. I suggest we rename our flags as well, for improved clarity.